src/Temporalio/Worker/WorkflowInstance.cs

```diff
@@ -491,13 +736,115 @@ public override void Init(WorkflowOutboundInterceptor outbound)
+                    {
+                        task = (Task)input.Definition.Delegate!.DynamicInvoke(input.Args)!;
+                    }
+                    await task.ConfigureAwait(true);
+                }
+                catch (TargetInvocationException e)
+                {
+                    ExceptionDispatchInfo.Capture(e.InnerException!).Throw();
+                }
+            }
+
+            /// <inheritdoc />
+            public override object? HandleQuery(HandleQueryInput input)
+            {
+                try
+                {
+                    if (input.Definition.Method is MethodInfo method)
+                    {
+                        return method.Invoke(instance.Instance, input.Args);
+                    }
+                    return input.Definition.Delegate!.DynamicInvoke(input.Args);
+                }
+                catch (TargetInvocationException e)
+                {
+                    ExceptionDispatchInfo.Capture(e.InnerException!).Throw();
+                    // Unreachable
+                    return null;
+                }
+            }
         }
 
         /// <summary>
         /// Workflow outbound implementation.
         /// </summary>
         internal class OutboundImpl : WorkflowOutboundInterceptor
         {
+            private readonly WorkflowInstance instance;
+
+            /// <summary>
+            /// Initializes a new instance of the <see cref="OutboundImpl"/> class.
+            /// </summary>
+            /// <param name="instance">Workflow instance.</param>
+            public OutboundImpl(WorkflowInstance instance) => this.instance = instance;
+
+            /// <inheritdoc />
+            public override Task DelayAsync(DelayAsyncInput input)
+            {
+                var token = input.CancellationToken ?? instance.CancellationToken;
+                // Task.Delay is started when created not when awaited, so we do the same. Also,
+                // they don't create the timer if already cancelled.
+                if (token.IsCancellationRequested)
+                {
+                    return Task.FromCanceled(token);
+                }
+
+                // After conferring, it was decided that < -1 is an error and 0 is 1ms and -1 can
+                // remain infinite (timer command never created)
+                var delay = input.Delay;
+                if (delay < TimeSpan.Zero && delay != Timeout.InfiniteTimeSpan)
+                {
+                    throw new ArgumentException("Delay duration cannot be less than 0.");
+                }
+                else if (delay == TimeSpan.Zero)
+                {
+                    delay = TimeSpan.FromMilliseconds(1);
+                }
+                var source = new TaskCompletionSource<object?>();
+                // Only create the command if not infinite. We use seq 0 to represent uncreated.
+                uint seq = 0;
+                if (delay != Timeout.InfiniteTimeSpan)
```

Comment:
Hmm.. only thing I'm worried about not creating the timer is that changing the duration from non infinite to infinite breaks determinism same way as it is in some other SDKs for changing from 0 to non 0, which we want to avoid.

Reply:
This is called out in the docs of `Workflow.Delay`:

```csharp
        /// The <c>delay</c> value can be <see cref="Timeout.Infinite" /> or
        /// <see cref="Timeout.InfiniteTimeSpan" /> but otherwise cannot be negative. A server-side
        /// timer is not created for infinite delays, so it is non-deterministic to change a timer
        /// to/from infinite from/to an actual value.
```

I think this is a reasonable tradeoff because in .NET people use infinite timers to create tasks that are just resolved on cancellation. Cancellation tokens don't really have a "task" you can wait on and we will probably forbid threaded `WaitHandle` things, so `await Workflow.DelayAsync(Timeout.Infinite, Workflow.CancellationToken)` may be a common pattern. We can try to avoid with helpers of course.

I could make a 100 year (or max duration or whatever) timer on the server, sure, but is it worth it just to allow that rare case of allowing people to change the timer duration later?

Reply:
Okay, let's keep it, it's well documented.
Some people will trip on this but I agree it's a reasonable tradeoff.